crates/bevy_input/src/gamepad.rs

```diff
@@ -1213,39 +1216,138 @@ impl AxisSettings {
     }
 
     /// Clamps the `raw_value` according to the `AxisSettings`.
-    pub fn clamp(&self, new_value: f32) -> f32 {
-        if self.deadzone_lowerbound <= new_value && new_value <= self.deadzone_upperbound {
+    pub fn clamp(&self, raw_value: f32) -> f32 {
+        if self.deadzone_lowerbound <= raw_value && raw_value <= self.deadzone_upperbound {
             0.0
-        } else if new_value >= self.livezone_upperbound {
+        } else if raw_value >= self.livezone_upperbound {
             1.0
-        } else if new_value <= self.livezone_lowerbound {
+        } else if raw_value <= self.livezone_lowerbound {
             -1.0
         } else {
-            new_value
+            raw_value
         }
     }
 
-    /// Determines whether the change from `old_value` to `new_value` should
+    /// Determines whether the change from `old_raw_value` to `new_raw_value` should
     /// be registered as a change, according to the [`AxisSettings`].
-    fn should_register_change(&self, new_value: f32, old_value: Option<f32>) -> bool {
-        if old_value.is_none() {
-            return true;
+    #[inline(always)]
+    fn should_register_change(&self, new_raw_value: f32, old_raw_value: Option<f32>) -> bool {
+        match old_raw_value {
+            None => true,
+            Some(old_raw_value) => ops::abs(new_raw_value - old_raw_value) >= self.threshold,
         }
-
-        ops::abs(new_value - old_value.unwrap()) > self.threshold
     }
 
-    /// Filters the `new_value` based on the `old_value`, according to the [`AxisSettings`].
+    /// Filters the `new_raw_value` based on the `old_raw_value`, according to the [`AxisSettings`].
     ///
-    /// Returns the clamped `new_value` if the change exceeds the settings threshold,
+    /// Returns the clamped and scaled `new_raw_value` if the change exceeds the settings threshold,
     /// and `None` otherwise.
-    pub fn filter(&self, new_value: f32, old_value: Option<f32>) -> Option<f32> {
-        let new_value = self.clamp(new_value);
+    #[inline(always)]
+    fn filter(
+        &self,
+        new_raw_value: f32,
+        old_raw_value: Option<f32>,
+    ) -> Option<FilteredAxisPosition> {
+        let clamped_unscaled = self.clamp(new_raw_value);
+        match self.should_register_change(clamped_unscaled, old_raw_value) {
+            true => Some(FilteredAxisPosition {
+                scaled: self.get_axis_position_from_value(clamped_unscaled),
+                raw: new_raw_value,
+            }),
+            false => None,
+        }
+    }
+
+    #[inline(always)]
+    fn get_axis_position_from_value(&self, value: f32) -> ScaledAxisWithDeadZonePosition {
+        if value < self.deadzone_upperbound && value > self.deadzone_lowerbound {
```

Comment:
Maybe use match? I think this way you won't need the `unreachable`.

Reply:
I tried but it didn't look better because it was all within guard clauses and it still needed an unreachable `_ => ...` branch. The compiler doesn't reason about guard clauses.

Reply:
Ah, I see :(